punchcut/src/source/glyf/outline.rs

```diff
@@ -0,0 +1,165 @@
+use super::Point;
+use crate::PathSink;
+
+/// TrueType outline.
+#[derive(Default, PartialEq, Eq, Debug)]
+pub struct Outline {
+    /// Set of points that define the shape of the outline.
+    pub points: Vec<Point>,
+    /// Set of tags (one per point).
+    pub tags: Vec<u8>,
+    /// Index of the end points for each contour in the outline.
+    pub contours: Vec<u16>,
+    /// True if the scaler applied a scale, in which case the points are in
+    /// 26.6 fixed point format. Otherwise, they are in integral font units.
+    pub is_scaled: bool,
+}
+
+impl Outline {
+    /// Creates a new empty outline.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Empties the outline.
+    pub fn clear(&mut self) {
+        self.points.clear();
+        self.tags.clear();
+        self.contours.clear();
+        self.is_scaled = false;
+    }
+
+    /// Converts the outline to a sequence of path commands and invokes the callback for
+    /// each on the given sink.
+    pub fn to_path(&self, sink: &mut impl PathSink) -> bool {
+        #[inline(always)]
+        fn conv(p: Point, s: f32) -> (f32, f32) {
+            (p.x as f32 * s, p.y as f32 * s)
+        }
+        const TAG_MASK: u8 = 0x3;
+        const CONIC: u8 = 0x0;
+        const ON: u8 = 0x1;
+        const CUBIC: u8 = 0x2;
+        let s = if self.is_scaled { 1. / 64. } else { 1. };
+        let points = &self.points;
+        let tags = &self.tags;
+        let mut count = 0usize;
+        let mut last_was_close = false;
+        for c in 0..self.contours.len() {
+            let mut cur = if c > 0 {
+                self.contours[c - 1] as usize + 1
+            } else {
+                0
+            };
+            let mut last = self.contours[c] as usize;
+            if last < cur || last >= points.len() {
+                return false;
+            }
+            let mut v_start = points[cur];
+            let v_last = points[last];
+            let mut tag = tags[cur] & TAG_MASK;
+            if tag == CUBIC {
+                return false;
+            }
+            let mut step_point = true;
+            if tag == CONIC {
```

Comment:
When do we encounter conics?

Reply:
This is actually a quadratic. The terminology was taken from FT_Outline_Decompose. I'll rename it.